exchanges/subscription/subscription.go

```diff
@@ -1,92 +1,163 @@
+		return k.String()
 	}
+	return fmt.Sprintf("%v: %s", key, ExactKey{s}.String())
+}
 
-	k := MaybePair{s.Enabled, s.Channel, s.Asset, s.Params, s.Interval, s.Levels, s.Authenticated, nil}
-	if s.Pair != currency.EMPTYPAIR {
-		k.Pair = &s.Pair
-	}
+// State returns the subscription state
+func (s *Subscription) State() State {
+	s.m.RLock()
+	defer s.m.RUnlock()
+	return s.state
+}
 
-	return json.Marshal(k)
+// SetState sets the subscription state
+// Errors if already in that state or the new state is not valid
+func (s *Subscription) SetState(state State) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	if state == s.state {
+		return ErrInStateAlready
+	}
+	if state > UnsubscribedState {
+		return ErrInvalidState
+	}
+	s.state = state
+	return nil
 }
 
-// String implements the Stringer interface for Subscription, giving a human representation of the subscription
-func (s *Subscription) String() string {
-	return fmt.Sprintf("%s %s %s", s.Channel, s.Asset, s.Pair)
+// SetKey does what it says on the tin safely for concurrency
+func (s *Subscription) SetKey(key any) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	s.Key = key
 }
 
-// EnsureKeyed sets the default key on a channel if it doesn't have one
-// Returns key for convenience
+// EnsureKeyed returns the subscription key
+// If no key exists then ExactKey will be used
 func (s *Subscription) EnsureKeyed() any {
-	if s.Key == nil {
-		s.Key = DefaultKey{
-			Channel: s.Channel,
-			Asset:   s.Asset,
-			Pair:    s.Pair,
-		}
+	// Juggle RLock/WLock to minimize concurrent bottleneck for hottest path
+	s.m.RLock()
+	if s.Key != nil {
+		defer s.m.RUnlock()
+		return s.Key
+	}
+	s.m.RUnlock()
+	s.m.Lock()
+	defer s.m.Unlock()
+	if s.Key == nil { // Ensure race hasn't updated Key whilst we swapped locks
+		s.Key = &ExactKey{s}
 	}
 	return s.Key
 }
+
+// Clone returns a copy of a subscription
+// Key is set to nil, because most Key types contain a pointer to the subscription, and because the clone isn't added to the store yet
+// Users should allow a default key to be assigned on AddSubscription or can SetKey as necessary
+func (s *Subscription) Clone() *Subscription {
+	s.m.RLock()
+	c := &Subscription{
+		Key:           nil,
+		Enabled:       s.Enabled,
+		Channel:       s.Channel,
+		Asset:         s.Asset,
+		Params:        s.Params,
+		Interval:      s.Interval,
+		Levels:        s.Levels,
+		Authenticated: s.Authenticated,
+		state:         s.state,
+		Pairs:         s.Pairs,
+	}
+	s.Pairs = slices.Clone(s.Pairs)
+	s.Params = maps.Clone(s.Params)
+	s.m.RUnlock()
+	return c
+}
+
+// SetPairs does what it says on the tin safely for concurrency
+func (s *Subscription) SetPairs(pairs currency.Pairs) {
+	s.m.Lock()
+	s.Pairs = pairs
+	s.m.Unlock()
+}
+
+// AddPairs does what it says on the tin safely for concurrency
+func (s *Subscription) AddPairs(pairs ...currency.Pair) {
+	if len(pairs) == 0 {
+		return
+	}
+	s.m.Lock()
+	for _, p := range pairs {
+		s.Pairs = s.Pairs.Add(p)
```

Comment:
out of scope: this method could be variadic as well. resolve for later up to you.

Reply:
Agreed, but I'll punt it. Thanks 👍